pkg/controller/clusterdeployment/metrics.go

```diff
@@ -83,12 +62,67 @@ func incProvisionFailedTerminal(cd *hivev1.ClusterDeployment) {
 	if stoppedCondition != nil {
 		stoppedReason = stoppedCondition.Reason
 	}
-	metricProvisionFailedTerminal.WithLabelValues(poolNSName,
-		hivemetrics.GetClusterDeploymentType(cd),
-		stoppedReason).Inc()
+	metricProvisionFailedTerminal.FixedLabels["clusterpool_namespacedname"] = poolNSName
+	metricProvisionFailedTerminal.FixedLabels["failure_reason"] = stoppedReason
```

Comment:
oo, this isn't good. The metric var, and its `FixedLabel` member, are global. These assignments can race, and we can end up observing with the wrong label values.

We either need to figure out a way to pass the fixed labels into `LogMetricWithDynamicLabels`, or curry them in advance. The former could be done by adding a `prometheus.Labels` param. If the currying could be done in one line, I would say it's sixes; but if we're still determined to avoid `MustCurryWith` and you want to check errors, let's do the former.

Reply:
Let me try to clarify:

Since `metricProvisionFailedTerminal` is a global variable, there's only one `metricsProvisionFailedTerminal.FixedLabels` map. Now imagine two threads reaching this code at the same time. (I'll use just the `clusterpool_namespacedname` fixed label to simplify the example.)
- Thread A: `FixedLabels["clusterpool_namespacedname"] = "foons/foopool"`
- Thread B: `FixedLabels["clusterpool_namespacedname"] = "barns/barpool"` <== overwrites the above <== **BAD**
- Thread B: `ObserveMetricWithDynamicLabels()` => observes with label value `"barns/barpool"`
- Thread A: `ObserveMetricWithDynamicLabels()` => *also* observes with label value `"barns/barpool"` <== **BAD**

What I'm suggesting to remedy this problem is one of the following.

### Option 1 ("pass the fixed labels in")
Change `ObserveMetricWithDynamicLabels` as follows:

```go
func (c CounterVecWithDynamicLabels) ObserveMetricWithDynamicLabels(obj metav1.Object, fixedLabels map[string]string, log log.FieldLogger, _ float64) {
	curriedVec, err := c.CounterVec.CurryWith(fixedLabels)
```

and this chunk becomes:

```go
	fixedLabels := map[string]string{
		"clusterpool_namespacedname": poolNSName,
		"failure_reason": stoppedReason,
	}
	metricProvisionFailedTerminal.ObserveMetricWithDynamicLabels(cd, fixedLabels, log, 1)
```

### ~~Option 2: ("curry them in advance")~~
(As I started to write this down, I realized it's Bad™ for a couple of reasons. Will explain below.)

Change this chunk to create a partially curried metric and call `ObserveMetricWithDynamicLabels` on that thing:
```go
	partiallyCurried, err := metricProvisionFailedTerminal.CurryWith(map[string]string{
		"clusterpool_namespacedname": poolNSName,
		"failure_reason": stoppedReason,
	})
	if err != nil {
		panic("!")
	}
	partiallyCurried.ObserveMetricWithDynamicLabels(cd, log, 1)
```

In addition to being more code than the above (since we have to error check both outside and inside the observer func), this doesn't work as written because the partially curried variable isn't of one of our `...WithDynamicLabels` types, so we would have to create one, and then how would we populate the `OptionalLabels`, blah blah messy gross. Or pass the partially curried var into the function, but now we've got two metric variables and have to be careful about which one to use, messy gross.

----

**TL;DR:** Go with Option 1.

That said, we still need to declare the fixed labels so we can register with them. So I suggest changing `FixedLabels` to either `[]string` or `sets.String` -- we only need the *keys* in the metric definition. 

Then if you like, you can add a helper function, called from each `ObserveMetricWithDynamicLabels` impl, that accepts the `fixedLabels` param passed in and validates that the keys therein are some kind of subset of the `FixedLabels` in the metric definition. Though that's probably overkill, since if they're not, `CurryWith` will error. Unless you want to enforce that `fixedLabels` contains *all* of the `FixedLabels`. Or fill in any missing ones with `MetricLabelDefaultValue`.

Does this make sense?

Reply:
That's a valid concern. I liked Option 1 as well and have implemented it to the best of my abilities, however many of such metrics still have empty fixedLabels and this can still happen. As a failsafe, one of the arguments to the Observe functions is v1 object - be that cluster deployment or provision instance. Is it safe to assume we shouldn't run into this problem now?